core/extraction/towns_fund_round_three.py

```diff
@@ -369,7 +369,7 @@ def extract_project_progress(df_data: pd.DataFrame, project_lookup: dict, round_
         try:
             # use nullable "Int64" type to allow for empty cells in the submission whilst retaining data as ints
             df_data[col] = df_data[col].astype("Int64")
-        except ValueError:
+        except (ValueError, TypeError):
```

Comment:
Are we definitely wanting to let non integer values through other than null? I may be inferring incorrectly here, but would it not be better to explicitly set conditions, rather than using a vague try/except catch-all?

So something like:
if cell is empty do blah..
if cell type is not in [x, y] then blah

Reply:
Is there an actual use case for having RAG as a float?

Reply:
if there is, then it should be explicitly set, as per the int, rather than doing it this way

Reply:
handling/catching exceptions in this way is often described as an Anti-pattern as all sorts of other unexpected errors can then slip through. In this case it is unlikely but is leaves the code vulnerable to bugs if the part within the try/except is ever extended.

Reply:
If a user copy & pastes a float value into a RAG rating column it will fall over because the try-except block that was already present in the code only looked for ValueErrors and not TypeErrors. We want it let it through transformation in order to be picked up as an enum validation failure, otherwise here it will just 500.

I feel like we want to let all TypeErrors through so they can be caught as enums failures and send back to the user rather than failing on this loop of code.

Reply:
ok got ya, so it does kind of follow the workflow as we deliberately let the error pass silently and then "kind of" re catch the error later in validation. 

So I had a look through, I can't see where it gets validated against an enum (validation schema round 3 has it as str). Am I missing something.

For me, this also raises the question, why do we cast it to int64 here at all? Why not just pass whatever value it has through (including null) and validate that against enum?

Reply:
```validate_enums``` used to work column-by-column, and so anything that could be converted to int64 was. Now it works cell by cell in order to a) retain the original row index, and b) because it's actually a more granular and accurate way of checking datatypes. This function needs to be refactored to take that into account. I think we can probably dispense with the anti-pattern of attempting to cast entire columns in a dtype in a try/catch block.

The RAG rating columns are all validated against the RagEnum enum:

```
"Delivery (RAG)": enums.RagEnum,
 "Spend (RAG)": enums.RagEnum,
 "Risk (RAG)": enums.RagEnum,
```

Reply:
All of the redundant code has been removed.